### ISSUE TITLE
useLazyQuery  Uncaught (in promise)DOMException: signal is aborted without reason

### ISSUE THREAD

**Issue by zouzhe66** [apollographql/apollo-client#10520]
###  **Uncaught (in promise)DOMException: signal is aborted without reason**
these below code maybe cause the exception，can we catch the exception？
 ```
abortControllersRef.current.forEach((controller) => {
        controller.abort();
      });
```
```
  asyncUpdate(signal: AbortSignal) {
    return new Promise<QueryResult<TData, TVariables>>((resolve, reject) => {
      const watchQueryOptions = this.watchQueryOptions;

      const handleAborted = () => {
        this.asyncResolveFns.delete(resolve)
        this.optionsToIgnoreOnce.delete(watchQueryOptions);
        signal.removeEventListener('abort', handleAborted)
        reject(signal.reason);
      };

      this.asyncResolveFns.add(resolve);
      this.optionsToIgnoreOnce.add(watchQueryOptions);
      signal.addEventListener('abort', handleAborted)
      this.forceUpdate();
    });
  }
```

**Comment by q234876317** [comment]
well done, thanks a lot!

**Comment by guicara** [comment]
Hello,

I have also the same issue after upgrading the `@apollo/client` library from version `3.7.1` to `3.7.8` when using `useLazyQuery` (probably the same with `useQuery`?).

The bug was introduced with version `3.7.4`.

**Comment by asgeo1** [comment]
I'm also having an issue here, and don't fully understand how this is marked resolved.

In my tests, using `@apollo/client` `3.7.9`, and React `18.0.0`, I get an `Uncaught (in promise)DOMException: signal is aborted without reason` error pretty much any time I use `useLazyQuery` in a component.

This seems to be because of React v18's `reactStrictMode`, which causes components in dev mode to always unmount and re-render. See: https://github.com/vercel/next.js/issues/35822

So I don't see how there's anyway to use `useLazyQuery` and `reactStrictMode: true` together, without triggering this error all the time ?

> A component that unmounts while a query is fetched truly is exceptional! 

I'm not sure, seems to happen all the time with latest React v18 and `reactStrictMode: true`

Surely we don't want to wrap everything in `try/catch` and check for `error.name === 'AbortError'` every time we use `useLazyQuery` ? It seems odd.

I know I could probably create my own version of `useLazyQuery` which by default ignores these errors (like @jerelmiller suggested here https://github.com/apollographql/apollo-client/pull/10427#issuecomment-1419679013)

So I'll either do that, or downgraded to `3.7.3`, as I don't want to disable `reactStrictMode`.

But just trying to understand whether this really is intended behaviour or not for React 18 and `reactStrictMode: true`

**Comment by 70nyIT** [comment]
agree with @asgeo1 . I have this issue too. It fires not when there's a query running, but every time I have a component with `useLazyQuery` implemented in it

**Comment by jerelmiller** [comment]
@asgeo1 I'm curious, how are you executing the query in strict mode? What you describe seems to imply that you're executing the query in a `useEffect`, which I'd argue is generally a code smell. You'd likely be better off using `useQuery` in that case.

---

That being said, there are some [legitimate cases](https://github.com/apollographql/apollo-client/issues/10533) for why this behavior is not ideal. I am considering a different approach that will instead let the promise resolve, but if the component is unmounted, the value will just disappear into the void (i.e. we won't try and update state). I think this would help in this situation.

**Comment by guicara** [comment]
@jerelmiller in our project we use `useLazyQuery` most of the time. it allows us to control when to trigger the query or mutation. In the following example, we have a custom hook to fetch some data. The `fetchUsers` method is then called in a `useEffect` hook, or when a user clicks on a button (for example).

```typescript
import { useLazyQuery } from '@apollo/client';
import { GET_USERS } from './query';

interface QueryData {
  users: User[];
}

interface QueryVars {
  sortBy: string;
  query: Pick<UserQueryInput, '_partition'>;
}

function useFetchUsers() {
  const [execute, { loading, error, data, refetch, called }] = useLazyQuery<QueryData, QueryVars>(
    GET_USERS,
    {
      notifyOnNetworkStatusChange: true,
    },
  );

  const fetchUsers = async (sortBy?: SortBy) => {
    const query: QueryVars = {
      query: { ... },
      sortBy,
    };

    const response = called
      ? await refetch(query)
      : await execute({
          variables: query,
        });

    return response.data?.users || [];
  };

  return { fetchUsers, loading, error, users: data?.users || [] };
}

export default useFetchUsers;
```

**Comment by clementAC** [comment]
I have the same problem in production so strict mode may not be the culprit. I have two types of AbortError:
- signal is aborted without reason
- The operation was aborted

**Comment by guicara** [comment]
For me, it's a bug AND a **breaking change**. A minor version upgrade should not cause something as important!

**Comment by 70nyIT** [comment]
I agree with @guicara . This is why I was asking the reason for this issue being market as "Closed"

**Comment by guicara** [comment]
What are the next steps @jerelmiller?

**Comment by jerelmiller** [comment]
I've changed my mind. I'm going to reopen this issue so that I can tag it in my PR. This will close again when the PR is merged.

**Comment by jerelmiller** [comment]
Hey all 👋 

We just released [v3.7.11](https://github.com/apollographql/apollo-client/releases/tag/v3.7.11) that contains a change to the behavior of `useLazyQuery` that will now allow the promise to resolve naturally rather than rejecting with an abort error when a component unmounts. I hope this works better for you!

**Comment by 70nyIT** [comment]
Confirmed working as expected! Thanks @jerelmiller
